provider/apple.go

```diff
@@ -0,0 +1,534 @@
+	jUser := r.FormValue("user") // json string with user name
+
+	oauthClaims, _, err := ah.JwtService.Get(r)
+	if err != nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusInternalServerError, err, "failed to get token")
+		return
+	}
+
+	if oauthClaims.Handshake == nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusForbidden, nil, "invalid handshake token")
+		return
+	}
+
+	retrievedState := oauthClaims.Handshake.State
+	if retrievedState == "" || retrievedState != state {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusForbidden, nil, "unexpected state")
+		return
+	}
+
+	var resp AppleVerificationResponse
+	err = ah.exchange(context.Background(), code, ah.makeRedirURL(r.URL.Path), &resp)
+	if err != nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusInternalServerError, err, "exchange failed")
+		return
+	}
+	ah.Logf("[DEBUG]response data %+v", resp)
+	if resp.Error != "" {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusInternalServerError, nil, fmt.Sprintf("token validation response error: %s", resp.Error))
+		return
+	}
+	// Get token claims for extract uid (and email or name if they exist in scope)
+	tokenClaims, err := ah.getTokenClaims(resp.IDToken)
+	if err != nil {
+		ah.L.Logf("failed to get claims: " + err.Error())
+		return
+	}
+
+	u := ah.mapUser(tokenClaims)
+
+	// try parse user name data if it exist in scope and service response
+	if jUser != "" {
+		ah.parseUserData(&u, jUser)
+	}
+
+	cid, err := randToken()
+	if err != nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusInternalServerError, err, "failed to make claim's id")
+		return
+	}
+	claims := token.Claims{
+		User: &u,
+		StandardClaims: jwt.StandardClaims{
+			Issuer:   ah.Issuer,
+			Id:       cid,
+			Audience: oauthClaims.Audience,
+		},
+		SessionOnly: false,
+	}
+
+	if _, err = ah.JwtService.Set(w, claims); err != nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusInternalServerError, err, "failed to set token")
+		return
+	}
+
+	ah.Logf("[DEBUG] user info %+v", u)
+
+	// redirect to back url if presented in login query params
+	if oauthClaims.Handshake != nil && oauthClaims.Handshake.From != "" {
+		http.Redirect(w, r, oauthClaims.Handshake.From, http.StatusTemporaryRedirect)
+		return
+	}
+	rest.RenderJSON(w, &u)
+
+}
+
+// LogoutHandler - GET /logout
+func (ah AppleHandler) LogoutHandler(w http.ResponseWriter, r *http.Request) {
+	if _, _, err := ah.JwtService.Get(r); err != nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusForbidden, err, "logout not allowed")
+		return
+	}
+	ah.JwtService.Reset(w)
+}
+
+// exchange sends the validation token request and gets access token and user claims
+// (e.g. https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens)
+func (ah *AppleHandler) exchange(ctx context.Context, code, redirectURI string, result interface{}) (err error) {
+
+	// check client_secret is expired and recreate new (client_secret JWT) if need
+	ok, err := ah.isClientSecretExpired()
```

Comment:
the whole handling with unverified makes me nervous as I can't see all possible attacks exploiting this. Is it possible to replace by the real verification?

Reply:
You are right! I will change it to verified parse.